File: algorithms/Compare Strings by Frequency of the Smallest Character/Compare Strings by Frequency of the Smallest Character.py
class Solution:
    def numSmallerByFrequency(self, queries: List[str], words: List[str]) -> List[int]:
        def _f(s):
            d = Counter(s)
            d =dict(sorted(d.items(), key=lambda item: item[0]))
            for x in d:
                return d[x]
        
        freq = []
        for w in words:
            n1 = _f(w)
            freq.append(n1)
        
        freq.sort(reverse=True)

        res = []
        for q in queries:
            n = _f(q)
            c=0
            for n1 in freq:
                if n < n1:
                    c+=1
                else:
                    break
            res.append(c)
        
        return res

File: scripts/algorithms/C/Compare Strings by Frequency of the Smallest Character/Compare Strings by Frequency of the Smallest Character.java
// Runtime: 5 ms (Top 87.83%) | Memory: 46.8 MB (Top 47.53%)
class Solution {
    public int[] numSmallerByFrequency(String[] queries, String[] words) {
        int[] ans = new int[queries.length];
        int[] freq = new int[words.length];
        for (int i = 0; i < words.length; i++) {
            freq[i] = freqOfSmallest(words[i]);
        }
        Arrays.sort(freq);
        int k = 0;
        for (String query : queries) {
            int target = freqOfSmallest(query);
            ans[k++] = binarySearch(freq, target);
        }
        return ans;
    }
    public int freqOfSmallest(String s) {
        int[] freq = new int[26];
        char min = 'z';
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            freq[c - 'a'] += 1;
            if (c < min) {
                min = c;
            }
        }
        return freq[min - 'a'];
    }
    public int binarySearch(int[] arr, int target) {
        int idx = arr.length;
        int lo = 0;
        int hi = idx - 1;
        int mid;
        while (lo <= hi) {
            mid = (lo + hi) / 2;
            if (arr[mid] <= target) {
                lo = mid + 1;
            } else {
                idx = mid;
                hi = mid - 1;
            }
        }
        return arr.length - idx;
    }
}